File: Implementierung/Backend/src/main/java/edu/kit/informatik/service/ParticipantService.java
package edu.kit.informatik.service;

import edu.kit.informatik.dto.mapper.interactions.ParticipantMapper;
import edu.kit.informatik.dto.userdata.interactions.ParticipantDto;
import edu.kit.informatik.exceptions.EntityNotFoundException;
import edu.kit.informatik.model.userdata.interactions.Participant;
import edu.kit.informatik.repositories.ParticipantRepository;
import org.springframework.transaction.annotation.Transactional;
import org.springframework.security.core.Authentication;
import org.springframework.security.oauth2.server.resource.authentication.JwtAuthenticationToken;
import org.springframework.stereotype.Component;

import java.util.List;
import java.util.Optional;

/**
 * Service für {@link Participant Teilnehmer}
 *
 * @author ugqbo
 * @version 1.0
 */

@Component
public class ParticipantService extends BaseService<Participant, ParticipantDto, ParticipantDto> {

    private static final String ID_ATTRIBUTE = "userId";

    private final ParticipantRepository participantRepository;

    /**
     * Konstruktor zum Erstellen eines Objektes der Klasse
     * @param participantRepository {@link ParticipantRepository}
     * @param participantMapper {@link ParticipantMapper}
     */
    public ParticipantService(ParticipantRepository participantRepository,
                              ParticipantMapper participantMapper) {
        super(participantMapper);
        this.participantRepository = participantRepository;
    }

    @Override
    public ParticipantDto add(ParticipantDto participantDto, Authentication authentication) {
        super.checkAuthorization(authentication, participantDto.getUserId());
        Participant newParticipant = this.participantRepository.save(this.mapper.dtoToModel(participantDto));

        return this.mapper.modelToDto(newParticipant);
    }

    @Transactional
    @Override
    public ParticipantDto update(ParticipantDto participantDto, Authentication authentication) {
        super.checkAuthorization(authentication, participantDto.getUserId());
        Optional<Participant> repositoryParticipantOptional = this.participantRepository
                                                                    .findParticipantById(participantDto.getId());

        Participant repositoryParticipant = repositoryParticipantOptional.orElseThrow(() ->
                                                new EntityNotFoundException(Participant.class, participantDto.getId()));
        Participant newParticipant = this.mapper.dtoToModel(participantDto);

        if (!newParticipant.getFirstName().equals(repositoryParticipant.getFirstName())) {
            repositoryParticipant.setFirstName(newParticipant.getFirstName());
            repositoryParticipant.setUpdatedAt(newParticipant.getUpdatedAt());
        }
        if (!newParticipant.getLastName().equals(repositoryParticipant.getLastName())) {
            repositoryParticipant.setLastName(newParticipant.getLastName());
            repositoryParticipant.setUpdatedAt(newParticipant.getUpdatedAt());
        }
        if (!newParticipant.getCourses().equals(repositoryParticipant.getCourses())) {
            repositoryParticipant.setCourses(newParticipant.getCourses());
            repositoryParticipant.setUpdatedAt(newParticipant.getUpdatedAt());
        }
        if (!newParticipant.getInteractions().equals(repositoryParticipant.getInteractions())) {
            repositoryParticipant.setInteractions(newParticipant.getInteractions());
            repositoryParticipant.setUpdatedAt(newParticipant.getUpdatedAt());
        }

        return mapper.modelToDto(repositoryParticipant);
    }

    @Override
    public ParticipantDto getById(String id, Authentication authentication) {
        Optional<Participant> participantOptional = this.participantRepository.findParticipantById(id);
        Participant participant = participantOptional.orElseThrow(() ->
                                        new EntityNotFoundException(Participant.class, id));
        super.checkAuthorization(authentication, participant.getUser().getId());

        return this.mapper.modelToDto(participant);
    }

    @Override
    public List<ParticipantDto> getAll(Authentication authentication) {
        JwtAuthenticationToken jAT = (JwtAuthenticationToken) authentication;

        return this.mapper.modelToDto(this.participantRepository.findParticipantsByUserId(
                                                                jAT.getTokenAttributes().get(ID_ATTRIBUTE).toString()));
    }

    @Transactional
    @Override
    public String delete(String id, Authentication authentication) {
        Optional<Participant> participantOptional = this.participantRepository.findParticipantById(id);
        Participant participant = participantOptional.orElseThrow(() ->
                                                                    new EntityNotFoundException(Participant.class, id));
        super.checkAuthorization(authentication, participant.getUser().getId());
        participant.setFirstName("Gelöschter");
        participant.setLastName("Teilnehmer");
        participant.setVisible(false);

        return id;
    }

    /**
     * Methode zum Löschen eines {@link Participant}
     * --> dient zum Löschen fpr andere Services
     * @param participant {@link Participant}
     */
    protected void delete(Participant participant) {
        participantRepository.deleteById(participant.getId());
    }
}

File: Implementierung/frontend/paschi/src/service/ParticipantService.ts
import {BASE_URL, BaseService} from "@/service/BaseService";
import { Participant } from "@/model/userdata/interactions/Participant";
import { ParticipantDto } from "@/dto/userdata/interactions/ParticipantDto";
import { ParticipantMapper } from "@/dto/mapper/interactions/ParticipantMapper";
import axios, { AxiosResponse } from "axios";
import { useUserStore } from "@/store/UserStore";

const PARTICIPANT_BASE_URL: string = BASE_URL + "/api/participant";

export class ParticipantService extends BaseService<Participant, ParticipantDto> {

  private static participantService: ParticipantService = new ParticipantService();

  private constructor() {
    super(ParticipantMapper.getMapper());
  }

  static getService(): ParticipantService {
    return this.participantService;
  }

  async add(participant: Participant) {
    const token = useUserStore().getUser()?.token;
    const participantDto = this.getMapper().modelToDto(participant);
    await axios
      .post(PARTICIPANT_BASE_URL, participantDto, {
        headers: {
          Authorization: `Bearer ${token}`,
        },
      })
      .then((response: AxiosResponse<ParticipantDto>) => {
        participant.setId = response.data.id;
      })
      .catch((error) => {
        console.log(error);
      });
  }

  async update(participant: Participant) {
    const token = useUserStore().getUser()?.token;
    const participantDto = this.getMapper().modelToDto(participant);
    await axios
      .put(PARTICIPANT_BASE_URL, participantDto, {
        headers: {
          Authorization: `Bearer ${token}`,
        },
      })
      .catch((error) => {
        console.log(error);
      });
  }

  async getById(id: string): Promise<Participant | undefined> {
    const token = useUserStore().getUser()?.token;
    let participant;
    await axios
      .get(PARTICIPANT_BASE_URL + `/${id}`, {
        headers: {
          Authorization: `Bearer ${token}`,
        },
      })
      .then(async (response: AxiosResponse<ParticipantDto>) => {
        participant = await this.getMapper().dtoToModel(response.data);
      })
      .catch((error) => {
        console.log(error);
      });

    if (participant != undefined) {
      return participant;
    } else {
      return undefined;
    }
  }

  async getAll(): Promise<Participant[]> {
    const token = useUserStore().getUser()?.token;
    let participants: Participant[] = [];
    await axios
      .get(PARTICIPANT_BASE_URL, {
        headers: {
          Authorization: `Bearer ${token}`,
        },
      })
      .then((response: AxiosResponse<ParticipantDto[]>) => {
        response.data.forEach(async (participantDto: ParticipantDto) => {
          participants.push(await this.getMapper().dtoToModel(participantDto));
        });
      })
      .catch((error) => {
        console.log(error);
      });

    return participants;
  }

  async delete(id: string) {
    const token = useUserStore().getUser()?.token;
    await axios
      .delete(PARTICIPANT_BASE_URL, {
        params: {
          id,
        },
        headers: {
          Authorization: `Bearer ${token}`,
        },
      })
      .catch((error) => {
        console.log(error);
      });
  }
}
